arborist/role.go

```diff
@@ -0,0 +1,222 @@
+package arborist
+
+import (
+	"errors"
+)
+
+// Representation of a role in the RBAC model.
+//
+// Subroles and permissions are sets of pointers to other roles and permissions.
+// (The `map[*Role]struct{}` etc. is just a hack to implement a "set" so we can
+// have constant-time lookup to check membership, as opposed to searching
+// through a slice.)
+type Role struct {
+	ID          string
+	Tags        map[string]struct{}
+	Permissions map[*Permission]struct{}
+	Subroles    map[*Role]struct{}
+
+	Parent *Role
+}
+
+// Create a new role with the given name and empty sets of subroles,
+// permissions, and tags.
+//
+// NOTE:
+//     - The new role does not point to a parent node yet.
+//     - The role ID is not guaranteed to be unique here; the engine must check
+//       that.
+func NewRole(ID string) (*Role, error) {
+	var role Role
+
+	role = Role{
+		ID:          ID,
+		Tags:        make(map[string]struct{}),
+		Permissions: make(map[*Permission]struct{}),
+		Subroles:    make(map[*Role]struct{}),
+		Parent:      nil,
+	}
+
+	return &role, nil
+}
+
+// Check for equality from one `Role` to another. Role names are enforced as
+// unique identifiers, so checking those for equality is sufficient.
+func (role *Role) equals(other_role *Role) bool {
+	return role.ID == other_role.ID
+}
+
+// Try to add a subrole underneath this role.
+//
+// This fails if there is already a subrole with the same name as the new one.
+//
+// NOTE: this function does not enforce global role name uniqueness, which the
+// `AuthEngine` handles. This function should not be used unless the engine has
+// already validated the new role to add.
+func (role *Role) insert(subrole *Role) error {
+	var err error
+
+	if _, contains := role.Subroles[subrole]; contains {
+		err = errors.New("role exists already")
+	} else {
+		role.Subroles[subrole] = struct{}{}
+	}
+
+	return err
+}
+
+// permit records that this role allows the given permission, adding it to the
+// set of permissions this role grants.
+func (role *Role) permit(permission *Permission) {
+	role.Permissions[permission] = struct{}{}
+}
+
+// hasTags checks if this role contains all the tags in the given list.
+func (role *Role) hasTags(tags []string) bool {
+	result := true
+	for _, tag := range tags {
+		if _, contains := role.Tags[tag]; !contains {
+			result = false
+			break
+		}
+	}
+	return result
+}
+
+// allSubroles returns a slice of all the subroles accessible in the tree
+// beneath this role.
+func (role *Role) allSubroles() []*Role {
```

Comment:
I like the first suggestion [here](https://stackoverflow.com/a/34466755) where you pass the consumer function

Reply:
unfortunately I think the use cases for this require recomputing it every time, and iterating through the whole thing, so it's not generator-like

Reply:
I would say it could be worth adding a method like that, but seems sort of anti-YAGNI

Reply:
You are still recomputing and iterating the fresh list everytime with the generator-consumer style?

Reply:
I meant, one of the nice use cases for a generator would be writing in a "lazy" fashion where the iteration can stop early once the caller finds what it needs; but there's no need for that for the roles because there's already a map for lookup